File: LinkedList/ReverseNodeInK-Group/reverseNode.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]):
        if head==None or head.next==None: return head
        newHead = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return newHead

    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right: return head
        n = 1
        temp = head
        a = None
        b = None
        c = None
        d = None
        while temp and n < right+1:
            if n==left-1:
                a = temp
            elif n==left:
                b = temp
            elif n==right:
                c = temp
            n += 1
            temp = temp.next
        d = c.next
        c.next = None
        t = self.reverseList(b)
        if a: a.next = t
        else: head = t
        b.next = d
        return head
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next: return head
        temp = head
        first = 1
        last = k
        t = temp
        while temp:
            remLen = 0
            while remLen < k and t:
                t=t.next
                remLen+=1
            if remLen == 0: break
            if remLen==k:
                temp = self.reverseBetween(temp,first,last)
                head = temp
            else: break
            first += k
            last += k
        return head

File: LinkedList/ReverseNodeInK-Group/reverseNode.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverse(ListNode* head){
        ListNode* Next = head;
        ListNode* curr = head;
        ListNode* prev = NULL;
        while(curr){
            Next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = Next;
        }
        return prev;
    }
    ListNode* reversePart(ListNode*  head,int left,int right){
        ListNode* a = NULL;
        ListNode* b = NULL;
        ListNode* c = NULL;
        ListNode* d = NULL;
        ListNode* temp = head;
        int n = 1;
        while(temp && n <= right){
            if(n == left-1) a = temp;
            else if(n == left) b = temp;
            else if(n == right) c = temp;

            temp = temp->next;
            n++;
        }
        d = c->next;
        c->next = NULL;
        ListNode* t = reverse(b);
        if (a) a->next = t;
        else head = t;
        b->next = d;
        return head;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(k==1)return head;
        if(!head || !head->next) return head;
        ListNode* temp = head;
        int first = 1;
        int last = k;
        ListNode* t = temp;
        while(temp){
            int remLen = 0;
            while(remLen < k && t){
                t=t->next;
                remLen+=1;
            }
            if(remLen == 0) break;
            if(remLen==k){
                temp = reversePart(temp,first,last);
                head = temp;
            }
            else break;
            first += k;
            last += k;
        }
        return head;
    }
};